components/operations-controller/controllers/operation_controller.go

```diff
@@ -16,36 +16,420 @@ package controllers
+	"github.com/kyma-incubator/compass/components/director/pkg/webhook"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/client"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/director"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/tenant"
+	web_hook "github.com/kyma-incubator/compass/components/operations-controller/internal/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"time"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+
+	corev1 "k8s.io/api/core/v1"
 
 	"github.com/go-logr/logr"
-	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 
-	operationsv1alpha1 "github.com/kyma-incubator/compass/components/operations-controller/api/v1alpha1"
+	"github.com/kyma-incubator/compass/components/operations-controller/api/v1alpha1"
 )
 
 // OperationReconciler reconciles a Operation object
 type OperationReconciler struct {
-	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
+	k8sClient      client.Client
+	webhookConfig  *web_hook.Config
+	directorClient director.Client
+	webhookClient  web_hook.Client
+	log            logr.Logger
+}
+
+func NewOperationReconciler(k8sClient client.Client, webhookConfig *web_hook.Config, directorClient director.Client, webhookClient web_hook.Client, logger logr.Logger) *OperationReconciler {
+	return &OperationReconciler{
+		k8sClient:      k8sClient,
+		webhookConfig:  webhookConfig,
+		log:            logger,
+		directorClient: directorClient,
+		webhookClient:  webhookClient,
+	}
 }
 
 // +kubebuilder:rbac:groups=operations.compass,resources=operations,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operations.compass,resources=operations/status,verbs=get;update;patch
 
 func (r *OperationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("operation", req.NamespacedName)
+	ctx := context.Background()
+	logger := r.log.WithValues("operation", req.NamespacedName)
+
+	operation, err := r.k8sClient.Get(ctx, req.NamespacedName)
+	if err != nil {
+		logger.Error(err, fmt.Sprintf("Unable to retrieve %s from API server", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	requestData, err := parseRequestData(operation)
+	if err != nil {
+		logger.Error(err, operationError("Unable to parse request data", operation))
+		return ctrl.Result{}, nil
+	}
+
+	ctx = tenant.SaveToContext(ctx, requestData.TenantID)
+	app, err := r.directorClient.FetchApplication(ctx, operation.Spec.ResourceID)
+	if err != nil {
+		logger.Error(err, operationError("Unable to fetch application", operation))
+
+		if isReconciliationTimeoutReached(operation, time.Duration(r.webhookConfig.TimeoutFactor)*r.webhookConfig.ReconciliationTimeout) {
+			if err := r.k8sClient.Delete(ctx, operation); err != nil {
```

Comment:
Firstly, the Director does not return an appropriate error message when app not found - it simply returns 200 OK and:

```
{
    "data": {
        "result": null
    }
}
```

Not sure it was intentionally left like so, because there were some old PRs which introduce more client-friendly errors in similar situations, but were not applied to the app not found case. Anyway, we'd have to do some changes in the Director to fix that, which is of course not something to sweat about.

 But it got me thinking in general, do we even need to interpret the error message? Why should we make the condition for deleting an Operation CR a combination of ROT expired + App Not Found? I think that the ROT expiration should dominate this decision and if it has passed, then there's nothing to try anymore with this operation and waste reconcile loop recourses.

Just image if there was some consistent non-app-not-found error which kept getting constantly returned - we'd never delete this Operation CR, because we'll never get the necessary App Not Found error.

What do you think?